### ISSUE TITLE
A question on only sax MRI prediction

### ISSUE THREAD

**Issue by ManZuRen** [ngaggion/HybridGNet_3D#3]
Hello! Thank you for this excellent work. I am currently using your method for mesh initialization in cardiac reconstruction. During usage, I encountered an issue (which may stem from insufficient understanding on my part): when making predictions using only SAX images (without LAX images), the cardiac mesh generated by your pre-trained model appears flattened along one axis. I wonder if this is due to the normalization of SAX images to the size 210×210×16? I look forward to your reply. Thank you!

<img width="504" height="815" alt="Image" src="https://github.com/user-attachments/assets/a12bc923-9f65-4156-8b73-d706ac39edae" />

<img width="771" height="828" alt="Image" src="https://github.com/user-attachments/assets/fb50a3b7-395a-489d-a1af-647fdbe0ddea" />

**Comment by ManZuRen** [comment]
I noticed that you have only released the pre-trained model for surface meshes. If I want to generate tetrahedral volume meshes, is it sufficient to just modify the configuration file (config)?

**Comment by ngaggion** [comment]
Hello! Yes, as we trained on UK Biobank, images were padded to 210x210x16 that was the biggest image size we had. So if the results look flattened, you should take a look into multiplying the output of the model by the voxelspacing, as the model will always produce the meshes in the image size (from 0, 0, 0, to 1, 1, 1) that when you multiply by image shape gets converted to the range (0, 0, 0) - (210, 210, 16). 
With respect to the volumetric image generation, I will upload the weights to [the huggingface dataset ](https://huggingface.co/datasets/ngaggion/HybridVNet_Weights/tree/main) today. On how to use them, please modify the inference example to load the volumetric info instead of the surface data.

**Comment by ManZuRen** [comment]
Thank you for your prompt feedback! I would also like to ask if the pre-trained model for tetrahedral meshes is only available for ROI_LAX. Actually, I want to generate an initial volume mesh using only SAX.

<img width="548" height="192" alt="Image" src="https://github.com/user-attachments/assets/e871f3bc-53d0-4bd0-97fc-a86921cca078" />

**Comment by ngaggion** [comment]
We do not have the trained model for other combinations, as on the paper we reported for the best performance that was doing the SAX+LAX with cropped SAX regions. I will take a look into it, to provide some other models in the future, at least for SAX only.

**Comment by ManZuRen** [comment]
Thank you for your response. I wonder if you could do me a favor: do you have any tetrahedral mesh files generated from your training (the ones mentioned in your paper that remain as a single integrated structure rather than being split into multiple parts)? Something like a template of a cardiac tetrahedral mesh. I would like to try using it as an initialization to see if it can optimize various personalized cardiac meshes.

**Comment by ManZuRen** [comment]
My English proficiency is quite limited, so many of my expressions may not be accurate. I’m also not sure if they are polite enough. Please bear with me.

**Comment by ngaggion** [comment]
Yes, indeed we have, [here](https://huggingface.co/datasets/ngaggion/HybridVNet_Weights/tree/main/VolumetricWeights/VolumetricFiles), that is a volumetrization of the Rodero et al whole heart atlas. We used that to generate our atlas dataset.
You have the tetrahedra defined as "vol_tets" and their points "vol_nodes". Using meshio you can create the mesh. Even have the points which belong to the surface as "surface_nodes_idx" and the faces of the surface as "surface_faces". Let me know if you need anything else.

**Comment by ManZuRen** [comment]
Hello, and sorry to bother you again. Thank you very much for your previous helpful reply.

I now have some data that match the input requirements of your model, namely SAX + 2CH + 4CH. However, my SAX stack contains only 14 slices along the z-axis, which does not meet the requirement of the pretrained model input size (210 × 210 × 16). I would like to ask whether you have any suggestions for handling this situation.

In addition, after using your model to generate the mesh, I would like to examine whether the mesh is properly aligned with the MRI by slicing the mesh and comparing the resulting cross-sections with the SAX images. I am not entirely sure how this alignment is handled in your method. More specifically, given one SAX slice, how can I determine the corresponding slicing position on the mesh so that the sliced mesh matches that particular MRI slice?

I would be very grateful for any advice you could share on these two questions.

Thank you again for your time and help.

**Comment by ngaggion** [comment]
Hello!

Please take a look at this inference code example I made to run some extra datasets.
https://github.com/ngaggion/HybridVNet/blob/main/inference_example.py

It makes use of these codes:

https://github.com/ngaggion/HybridVNet/blob/main/utils/dataset_inference_w_LAX_example.py
This dataset inference class pads the images centered on the z-axis. so if you have less slices (which we do), you pad.
Look at the class PadArraysToSquareShape(object) method to see how this works.

https://github.com/ngaggion/HybridVNet/blob/main/utils/inference_aux.py
This code contains the go_back function. As we are working on the sax image space, and all positions are relative to the internal image space, we need to deal with:
1- the padding we may have done
2- the scaling and the voxelspacing
3- any rotations from the physical plane
4- incorporating the origin.

The method itself is completely agnostic to the physical space your images will live in. 
We obtain all the information required by looking at the metadata using simpleitk to load the images.

Sorry if it is a bit messy, but we have dealt with many institutions storing the data in different ways and formats, and was almost impossible to get a straightforward way to the original space. We use meshio to store stl in order to open it with visualization tools as MITK

**Comment by ManZuRen** [comment]
Hello,

Thank you very much! I think I was able to use your code to do the padding and align the generated mesh back to the real SAX space. However, I am still seeing an issue in the last few slices of the SAX sequence.

I rendered the cross-sections of the tetrahedral mesh and overlaid them on the real SAX MRI images. The shape looks correct in the middle slices, although there may be some small offsets. But in the last few slices of the sequence, the alignment becomes incorrect. My impression is that, in those slices, we seem to be cutting through the “cardiac vessel” part of the mesh, while the real image is still showing the left and right ventricles.

Have you encountered this issue before?

Thank you very much for your help.

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/621762de-bea5-437d-8ab1-161d87de6bf3" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/1ecdfc4e-df32-4246-94a2-75cc1d49c10a" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/2d1e9204-f6e4-417f-9eb7-eaf8ef24ca6e" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/610a9033-ceac-44ba-a41c-eb708a48677c" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/10596673-3293-46c9-b609-6af9b2683ede" />

**Comment by ngaggion** [comment]
Are you using your own data? I'm wondering what could be the issue.
Either it can be that the model does not perform that well outside UKBiobank, or that you may have to flip the z-axis before putting it as input to the model. If what you are saying is that it only works correctly in the middle slices

**Comment by ManZuRen** [comment]
I suspect the issue may come from my dataset itself. Its SAX stack seems to be acquired more toward the apex, so if I pad one blank slice on both sides of the SAX sequence, the generated mesh tends to be biased toward the apex overall. I tried padding two blank slices only at the end of the SAX sequence, and the result seems to be better now.

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/b0466c2c-c04b-42d1-8c1f-2a349018158b" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/88332059-5c3a-4b60-be5e-7672e837d260" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/12054db2-08b8-4216-8c5a-3d62de59b942" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/eb11541c-6b0a-4688-8bcf-eb54e2e9541e" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/321067e7-761b-41e3-87d5-0ee1478f006e" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/d2bedbdd-8046-4d6e-8efe-3b8cd76ad644" />

I’m not sure whether this is really the reason. The UK Biobank data is quite difficult for me to access, so I used an older dataset instead — the Cardiac Atlas Project 2011. After segmenting it and rendering it volumetrically, I obtained something like the following. I’m not sure whether the issue comes from the fact that its slice coverage is closer to the apex, which may cause the predicted mesh to be less accurate around the region where the heart connects to the vessels.

<img width="359" height="333" alt="Image" src="https://github.com/user-attachments/assets/60956d1b-dd70-45f4-bbdb-242f34355f4d" />

Or do you have any easily accessible datasets that you would recommend for this project?
Thank you very much for your prompt reply！

**Comment by ngaggion** [comment]
Indeed, I finished my collaboration with UK so I also do not have access to uk biobank now, and I'm also no longer working on this topic so sorry but I won't be able to help you more than this. 
Your meshes look quite nice, remember you are using different data from a different source, so results won't have the same quality as images taken with the same protocol

**Comment by ManZuRen** [comment]
Thank you very much for your reply. I’ve decided to use this result. Your response was very timely and truly helpful to me. You are a very responsible author. Thank you again for your help!
